core/src/main/java/org/geysermc/geyser/skin/SkinProvider.java

```diff
@@ -260,42 +259,33 @@ static CompletableFuture<SkinProvider.SkinData> requestSkinData(PlayerEntity ent
                             cape = getCachedBedrockCape(entity.getUuid());
                         }
 
-                        if (cape.failed() && ALLOW_THIRD_PARTY_CAPES) {
-                            cape = getOrDefault(requestUnofficialCape(
-                                    cape, entity.getUuid(),
-                                    entity.getUsername(), false
-                            ), EMPTY_CAPE, CapeProvider.VALUES.length * 3);
-                        }
+                        // Call event to allow extensions to modify the skin, cape and geo
+                        boolean isBedrock = GeyserImpl.getInstance().connectionByUuid(entity.getUuid()) != null;
+                        SkinData skinData = new SkinData(skin, cape, geometry);
+                        final EventSkinData eventSkinData = new EventSkinData(skinData);
+                        GeyserImpl.getInstance().eventBus().fire(new SessionSkinApplyEvent(session, entity.getUsername(), entity.getUuid(), data.isAlex(), isBedrock, skinData) {
+                            @Override
+                            public SkinData skinData() {
+                                return eventSkinData.skinData();
+                            }
 
-                        boolean isDeadmau5 = "deadmau5".equals(entity.getUsername());
-                        // Not a bedrock player check for ears
-                        if (geometry.failed() && (ALLOW_THIRD_PARTY_EARS || isDeadmau5)) {
-                            boolean isEars;
-
-                            // Its deadmau5, gotta support his skin :)
-                            if (isDeadmau5) {
-                                isEars = true;
-                            } else {
-                                // Get the ears texture for the player
-                                skin = getOrDefault(requestUnofficialEars(
-                                        skin, entity.getUuid(), entity.getUsername(), false
-                                ), skin, 3);
-
-                                isEars = skin.isEars();
+                            @Override
+                            public void skin(@NonNull Skin newSkin) {
+                                eventSkinData.skinData(new SkinData(newSkin, skinData.cape(), skinData.geometry()));
                             }
```

Comment:
I'd add a `Objects.requireNonNull(newSkin)` here and on all other api methods with a `@NonNull` annotation - otherwise, it's possible to pass a null value